docs/TrafficSplitting.md

```diff
@@ -0,0 +1,163 @@
+- If a Revision is untagged and it has 0% traffic split, its removed from the traffic block entirely.
+
+## Flag operation precedence
+- Mentioned 5 flags for traffic splitting and tagging operations can be specified together in a single `kn service update` command, thus `kn` defines the precedence of these flags in which operations are evaluated.
+- The order of the flags specified in the command are **not** taken into account.
+- Enlisting the precedence of flag(s) as they are evaluated by `kn`:
+
+1. `--untag-revision` : First, all the referenced tags by this flag are removed from the traffic block
+2. `--tag-revision` or `--tag-latest` : Second, all the revisions are tagged as specified in the traffic block
+3. `--traffic` or `--traffic-latest`: Third, all the referenced revisions (or tags) are assigned given portion of traffic split
+
+Thus, if a user wants to update tag for revision `echo-v2` from `new` to `old` with `50%` traffic while adjusting rest to latest revision, will do
+```
+kn service update svc --untag-revision new --tag-revision echo-v2:old --traffic old:50 --traffic-latest 50
+```
+
+## Summary
+Following table shows quick summary for traffic splitting flags, their value formats, operation to perform and *Repetition* column denotes if repeating the flag is allowed in a `kn service update` command.
+
+
+Flag | Value(s) | Operation | Repetition
+:--- | :---: | :--- | :---:
+`--traffic` | `RevisionName:Percent` | Give `Percent` traffic to `RevisionName` | :heavy_check_mark:
+`--traffic` | `Tag:Percent` | Give `Percent` traffic to the Revision having `Tag` | :heavy_check_mark:
+`--traffic-latest` | `Percent` | Give `Percent` traffic to the latest ready Revision | :heavy_multiplication_x:
+`--tag-revision` | `RevisionName:Tag` | Give `Tag` to `RevisionName` | :heavy_check_mark:
+`--tag-latest` | `Tag` | Give `Tag` to the latest ready Revision | :heavy_multiplication_x:
+`--untag-revision` | `Tag` | Remove `Tag` from Revision | :heavy_check_mark:
+
+
+## Examples
+1. Tag revisions `echo-v1` and `echo-v2` as `stable` and `staging` respectively:
+```
+kn service update svc --tag-revision echo-v1:stable --tag-revision echo-v2:staging
+```
+
+2. Ramp up/down revision `echo-v3` to `20%`, adjusting other traffic to accommodate:
+```
+kn service update svc --traffic echo-v3:20 --traffic echo-v2:80
+```
+
+3. Give revision `echo-v3` the tag `candidate`, without otherwise changing any traffic split:
+```
+kn service update svc --tag-revision echo-v3:candidate
+```
+
+4. Give `echo-v3` the tag `candidate`, and `2%` of traffic adjusting other traffic to go to revision `echo-v2`:
+```
+kn service update svc --tag-revision echo-v3:candidate --traffic candidate:2 --traffic echo-v2:98
+```
+
+6. Update the tag for `echo-v3` from `candidate` to `current`:
+```
+kn service update svc --untag-revision echo-v3 --tag-revision echo-v3:current
```

Comment:
Can we disallow to overwrite the tag to the same revision?

```bash
# Create `candidate` tag for `echo-v3` revision
kn service update svc --tag-revision echo-v3:candidate
# Update the tag for `echo-v3` from `candidate` to `current`
kn service update svc --tag-revision echo-v3:current
```

Reply:
> Can we disallow to overwrite the tag to the same revision?

We don't allow overwriting the tag to the same revision.
- If specified a **new** tag for **same** revision which is part of **existing** traffic block, this adds a new target in traffic block with 0% percent traffic split (if traffic isn't specified, else do as mentioned).
- If the mentioned tag **present** in **existing** traffic block for **different** revision, it raises an error.

The reason we've settled on the single entry point for traffic splitting is to be able to specify the desired state in single `service update` command. 

> Create `candidate` tag for `echo-v3` revision
> kn service update svc --tag-revision echo-v3:candidate
> Update the tag for `echo-v3` from `candidate` to `current`
> kn service update svc --tag-revision echo-v3:current
 
What happens if `echo-v3` is tagged twice in existing service traffic block ? 
(we can't uniquely identify a target in traffic block by just referencing revision name)

Tagging is a tricky operation with variants as
1. Give a new Tag 'new' to revision 'echo-v2'
```
--tag-revision echo-v2:new
```
2. Update Tag to 'old' from 'new' ' for revision echo-v2 and give Tag 'new' to `echo-v3` 
```
--untag-revision new --tag-revision echo-v2:old --tag-revision echo-v3:new
```
3. Duplicate revision echo-v3 tags as 'new' and 'current' with 50-50 traffic split
```
--tag-revision echo-v3:current --traffic new:50 --traffic current:50
# sees that echo-v3 is already tagged as 'new' in existing traffic block, thus adds a new target
# with tag 'current' for echo-v3 and assigns 50% traffic to it, while also updating traffic for
# 'new' to 50%
```

@toVersus : does this make sense?

Reply:
@navidshaikh Thanks for writing it up! Your detailed explanation really helped.